File: src/test/java/org/railwaystations/rsapi/core/services/ProfileServiceTest.java
package org.railwaystations.rsapi.core.services;

import org.junit.jupiter.api.Test;
import org.railwaystations.rsapi.adapter.out.db.OAuth2AuthorizationDao;
import org.railwaystations.rsapi.adapter.out.db.UserDao;
import org.railwaystations.rsapi.adapter.out.monitoring.MockMonitor;
import org.railwaystations.rsapi.app.auth.LazySodiumPasswordEncoder;
import org.railwaystations.rsapi.app.config.MessageSourceConfig;
import org.railwaystations.rsapi.core.model.License;
import org.railwaystations.rsapi.core.model.User;
import org.railwaystations.rsapi.core.ports.in.ManageProfileUseCase;
import org.railwaystations.rsapi.core.ports.out.Mailer;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.matches;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;
import static org.railwaystations.rsapi.adapter.in.web.controller.ProfileControllerTest.EXISTING_USER_ID;
import static org.railwaystations.rsapi.adapter.in.web.controller.ProfileControllerTest.USER_EMAIL;
import static org.railwaystations.rsapi.adapter.in.web.controller.ProfileControllerTest.USER_NAME;

class ProfileServiceTest {

    public static final String USER_AGENT = "UserAgent";

    private final UserDao userDao = mock(UserDao.class);

    private final MockMonitor monitor = new MockMonitor();

    private final Mailer mailer = mock(Mailer.class);

    private final OAuth2AuthorizationDao authorizationDao = mock(OAuth2AuthorizationDao.class);

    private final ProfileService sut = new ProfileService(monitor, mailer, userDao, authorizationDao, "EMAIL_VERIFICATION_URL", new LazySodiumPasswordEncoder(), new MessageSourceConfig().messageSource());

    @Test
    void testRegisterInvalidData() {
        var newUser = User.builder()
                .name("nickname")
                .ownPhotos(true)
                .build();
        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void registerNewUser() {
        var newUser = createNewUser().build();

        sut.register(newUser, USER_AGENT);

        verify(userDao).findByNormalizedName(USER_NAME);
        verify(userDao).countBlockedUsername(USER_NAME);
        verify(userDao).findByEmail(USER_EMAIL);
        verify(userDao).insert(any(User.class), anyString(), anyString());
        verify(userDao, never()).updateCredentials(anyInt(), anyString());

        assertThat(monitor.getMessages().getFirst()).isEqualTo("New registration{nickname='%s', email='%s'}\nvia %s".formatted(USER_NAME, USER_EMAIL, USER_AGENT));
        assertNewPasswordEmail();

        verifyNoMoreInteractions(userDao);
    }


    public static void assertVerificationEmail(Mailer mailer) {
        verify(mailer, times(1))
                .send(anyString(),
                        anyString(), matches("""
                                Hello,

                                please click on EMAIL_VERIFICATION_URL.* to verify your eMail-Address.

                                Cheers
                                Your Railway-Stations-Team

                                ---
                                Hallo,

                                bitte klicke auf EMAIL_VERIFICATION_URL.*, um Deine eMail-Adresse zu verifizieren.

                                Viele Grüße
                                Dein Bahnhofsfoto-Team"""));
    }

    @Test
    void registerNewUserWithPassword() {
        var newUser = createNewUser()
                .newPassword("verySecretPassword")
                .build();

        sut.register(newUser, USER_AGENT);

        verify(userDao).findByNormalizedName(USER_NAME);
        verify(userDao).countBlockedUsername(USER_NAME);
        verify(userDao).findByEmail(USER_EMAIL);
        verify(userDao).insert(any(User.class), anyString(), anyString());
        verify(userDao, never()).updateCredentials(anyInt(), anyString());

        assertThat(monitor.getMessages().getFirst()).isEqualTo("New registration{nickname='%s', email='%s'}\nvia %s".formatted(USER_NAME, USER_EMAIL, USER_AGENT));
        assertVerificationEmail(mailer);

        verifyNoMoreInteractions(userDao);
    }

    @Test
    void updateMyProfileNewMail() {
        when(userDao.findByEmail("newname@example.com")).thenReturn(Optional.empty());
        var existingUser = givenExistingUser();
        var updatedUser = createNewUser().email("newname@example.com").build();

        sut.updateProfile(existingUser, updatedUser, USER_AGENT);

        assertVerificationEmail(mailer);
        var user = User.builder()
                .id(EXISTING_USER_ID)
                .name(USER_NAME)
                .license(License.CC0_10)
                .email("newname@example.com")
                .ownPhotos(true)
                .anonymous(false)
                .url("http://twitter.com/")
                .sendNotifications(true).build();
        verify(userDao).update(user.getId(), user);
    }

    @Test
    void changePasswordTooShortBody() {
        var user = givenExistingUser();

        assertThatThrownBy(() -> sut.changePassword(user, "secret"))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void changePasswordBody() {
        var user = givenExistingUser();

        sut.changePassword(user, "secretlong");

        verify(userDao).updateCredentials(eq(user.getId()), anyString());
        verify(authorizationDao).deleteAllByUser(user.getName());
    }

    @Test
    void verifyEmailSuccess() {
        var token = "verification";
        var user = User.builder()
                .id(EXISTING_USER_ID)
                .name(USER_NAME)
                .license(License.CC0_10)
                .email(USER_EMAIL)
                .ownPhotos(true)
                .anonymous(false)
                .url("https://link@example.com")
                .emailVerification(token)
                .sendNotifications(true).build();
        when(userDao.findByEmailVerification(token)).thenReturn(Optional.of(user));

        sut.emailVerification(token);

        assertThat(monitor.getMessages().getFirst()).isEqualTo(String.format("Email verified {nickname='%s', email='%s'}", USER_NAME, USER_EMAIL));
        verify(userDao).updateEmailVerification(EXISTING_USER_ID, User.EMAIL_VERIFIED);
    }

    @Test
    void verifyEmailFailed() {
        var token = "verification";
        var user = User.builder()
                .id(EXISTING_USER_ID)
                .name(USER_NAME)
                .license(License.CC0_10)
                .email(USER_EMAIL)
                .ownPhotos(true)
                .anonymous(false)
                .url("https://link@example.com")
                .emailVerification(token)
                .sendNotifications(true).build();
        when(userDao.findByEmailVerification(token)).thenReturn(Optional.of(user));

        sut.emailVerification("wrong_token");

        assertThat(monitor.getMessages().isEmpty()).isTrue();
        verify(userDao, never()).updateEmailVerification(EXISTING_USER_ID, User.EMAIL_VERIFIED);
    }

    @Test
    void resendEmailVerification() {
        var user = givenExistingUser();

        sut.resendEmailVerification(user);

        assertVerificationEmail(mailer);
        verify(userDao).updateEmailVerification(eq(EXISTING_USER_ID), anyString());
    }

    @Test
    void deleteMyProfile() {
        var user = givenExistingUser();

        sut.deleteProfile(user, USER_AGENT);

        verify(userDao).anonymizeUser(eq(EXISTING_USER_ID));
        verify(userDao).addUsernameToBlocklist(USER_NAME);
        verify(authorizationDao).deleteAllByUser(USER_NAME);
    }

    @Test
    void registerUserNameTaken() {
        var user = givenExistingUser();
        var newUser = createNewUser()
                .name(user.getName())
                .build();

        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(ManageProfileUseCase.ProfileConflictException.class);
    }

    @Test
    void registerUserNameBlocked() {
        var newUser = createNewUser().name("Blocked Name").build();
        when(userDao.countBlockedUsername("blockedname")).thenReturn(1);

        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(ManageProfileUseCase.ProfileConflictException.class);
    }

    @Test
    void registerUserEmailTaken() {
        var user = givenExistingUser();
        var newUser = createNewUser()
                .name("othername")
                .email(user.getEmail())
                .build();

        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(ManageProfileUseCase.ProfileConflictException.class);

        assertThat(monitor.getMessages().getFirst()).isEqualTo("Registration for user 'othername' with eMail 'existing@example.com' failed, eMail is already taken\nvia UserAgent");
    }

    @Test
    void registernUserNameTaken() {
        var user = givenExistingUser();
        var newUser = createNewUser()
                .name(user.getName())
                .email("otheremail@example.com")
                .build();

        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(ManageProfileUseCase.ProfileConflictException.class);

        assertThat(monitor.getMessages().getFirst()).isEqualTo("Registration for user '%s' with eMail '%s' failed, name is already taken by different eMail '%s'%nvia %s"
                .formatted(user.getName(), newUser.getEmail(), user.getEmail(), USER_AGENT));
    }

    @Test
    void registerUserWithEmptyName() {
        var newUser = createNewUser()
                .name("")
                .build();

        assertThatThrownBy(() -> sut.register(newUser, USER_AGENT))
                .isInstanceOf(IllegalArgumentException.class);
    }

    private void assertNewPasswordEmail() {
        verify(mailer, times(1))
                .send(anyString(),
                        anyString(), matches("""
                                Hello,
                                                    
                                your new password is: .*
                                                    
                                Cheers
                                Your Railway-Stations-Team"""));
    }

    private User.UserBuilder createNewUser() {
        var key = "246172676F6E32696424763D3139246D3D36353533362C743D322C703D3124426D4F637165757646794E44754132726B566A6A3177246A7568362F6E6C2F49437A4B475570446E6B674171754A304F7A486A62694F587442542F2B62584D49476300000000000000000000000000000000000000000000000000000000000000";
        return User.builder()
                .name(USER_NAME)
                .email(USER_EMAIL)
                .url("https://link@example.com")
                .license(License.CC0_10)
                .ownPhotos(true)
                .anonymous(false)
                .key(key)
                .admin(false)
                .sendNotifications(true);
    }

    private User givenExistingUser() {
        var user = createNewUser()
                .id(EXISTING_USER_ID)
                .build();
        when(userDao.findByEmail(user.getEmail())).thenReturn(Optional.of(user));
        when(userDao.findByNormalizedName(user.getName())).thenReturn(Optional.of(user));
        return user;
    }

    @Test
    void resetPasswordUnknownUser() {
        assertThatThrownBy(() -> sut.resetPassword("unknown_user", USER_AGENT))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void resetPasswordEmptyEmail() {
        when(userDao.findByNormalizedName("nickname")).thenReturn(Optional.of(User.builder().build()));

        assertThatThrownBy(() -> sut.resetPassword("nickname", USER_AGENT))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void resetPasswordViaUsernameEmailNotVerified() {
        when(userDao.findByNormalizedName("nickname")).thenReturn(Optional.of(User.builder().id(123).name("nickname").email("nickname@example.com").build()));

        sut.resetPassword("nickname", USER_AGENT);

        verify(userDao).updateCredentials(eq(123), anyString());
        assertThat(monitor.getMessages().getFirst()).isEqualTo("Reset Password for 'nickname', email='nickname@example.com'");
        assertNewPasswordEmail();
        verify(userDao).updateEmailVerification(123, User.EMAIL_VERIFIED_AT_NEXT_LOGIN);
        verify(authorizationDao).deleteAllByUser("nickname");
    }

    @Test
    void resetPasswordViaEmailAndEmailVerified() {
        when(userDao.findByEmail("nickname@example.com")).thenReturn(Optional.of(User.builder().id(123).name("nickname").email("nickname@example.com").emailVerification(User.EMAIL_VERIFIED).build()));

        sut.resetPassword("nickname@example.com", USER_AGENT);

        verify(userDao).updateCredentials(eq(123), anyString());
        assertThat(monitor.getMessages().getFirst()).isEqualTo("Reset Password for 'nickname', email='nickname@example.com'");
        assertNewPasswordEmail();
        verify(userDao, never()).updateEmailVerification(123, User.EMAIL_VERIFIED_AT_NEXT_LOGIN);
        verify(authorizationDao).deleteAllByUser("nickname");
    }

}

File: core/src/test/kotlin/org/railwaystations/rsapi/core/services/ProfileServiceTest.kt
package org.railwaystations.rsapi.core.services

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.railwaystations.rsapi.core.config.MessageSourceConfig
import org.railwaystations.rsapi.core.model.License
import org.railwaystations.rsapi.core.model.User
import org.railwaystations.rsapi.core.model.UserTestFixtures.EXISTING_USER_ID
import org.railwaystations.rsapi.core.model.UserTestFixtures.USER_AGENT
import org.railwaystations.rsapi.core.model.UserTestFixtures.USER_EMAIL
import org.railwaystations.rsapi.core.model.UserTestFixtures.USER_NAME
import org.railwaystations.rsapi.core.ports.inbound.ManageProfileUseCase.ProfileConflictException
import org.railwaystations.rsapi.core.ports.outbound.MailerPort
import org.railwaystations.rsapi.core.ports.outbound.MonitorPort
import org.railwaystations.rsapi.core.ports.outbound.OAuth2AuthorizationPort
import org.railwaystations.rsapi.core.ports.outbound.UserPort
import org.springframework.security.crypto.password.PasswordEncoder

internal class ProfileServiceTest {
    private val userPort = mockk<UserPort>()

    private val monitorPort = mockk<MonitorPort>()

    private val mailerPort = mockk<MailerPort>(relaxed = true)

    private val authorizationPort = mockk<OAuth2AuthorizationPort>(relaxed = true)

    private val sut = ProfileService(
        monitorPort = monitorPort,
        mailerPort = mailerPort,
        userPort = userPort,
        authorizationPort = authorizationPort,
        eMailVerificationUrl = "EMAIL_VERIFICATION_URL",
        passwordEncoder = mockk<PasswordEncoder>(relaxed = true),
        messageSource = MessageSourceConfig().messageSource()
    )

    @BeforeEach
    fun setup() {
        every { userPort.addUsernameToBlocklist(any()) } returns Unit
        every { userPort.anonymizeUser(any()) } returns Unit
        every { userPort.countBlockedUsername(any()) } returns 0
        every { userPort.findByEmail(any()) } returns null
        every { userPort.findByNormalizedName(any()) } returns null
        every { userPort.findByEmailVerification(any()) } returns null
        every { userPort.insert(any(), any(), any()) } returns 0
        every { userPort.update(any(), any()) } returns Unit
        every { userPort.updateCredentials(any(), any()) } returns Unit
        every { userPort.updateEmailVerification(any(), any()) } returns Unit
        every { monitorPort.sendMessage(any()) } returns Unit
    }

    @Test
    fun testRegisterInvalidData() {
        val newUser = User(
            name = "nickname",
            license = License.CC0_10,
            ownPhotos = true,
        )
        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun registerNewUser() {
        val newUser = createNewUser()

        sut.register(newUser, USER_AGENT)

        verify { userPort.findByNormalizedName(USER_NAME) }
        verify { userPort.countBlockedUsername(USER_NAME) }
        verify { userPort.findByEmail(USER_EMAIL) }
        verify { userPort.insert(any(), any(), any()) }
        verify(exactly = 0) { userPort.updateCredentials(any(), any()) }
        verify {
            monitorPort.sendMessage(
                "New registration{nickname='%s', email='%s'}\nvia %s".format(
                    USER_NAME, USER_EMAIL, USER_AGENT
                )
            )
        }
        assertNewPasswordEmail()
    }


    @Test
    fun registerNewUserWithPassword() {
        val newUser = createNewUser().copy(
            newPassword = "verySecretPassword"
        )

        sut.register(newUser, USER_AGENT)

        verify { userPort.findByNormalizedName(USER_NAME) }
        verify { userPort.countBlockedUsername(USER_NAME) }
        verify { userPort.findByEmail(USER_EMAIL) }
        verify { userPort.insert(any(), any(), any()) }
        verify(exactly = 0) { userPort.updateCredentials(any(), any()) }
        verify {
            monitorPort.sendMessage(
                "New registration{nickname='%s', email='%s'}\nvia %s".format(
                    USER_NAME, USER_EMAIL, USER_AGENT
                )
            )
        }
        assertVerificationEmail(mailerPort)
    }

    @Test
    fun updateMyProfileNewMail() {
        every { userPort.findByEmail("newname@example.com") } returns null
        val existingUser = givenExistingUser()
        val updatedUser = createNewUser().copy(
            id = existingUser.id,
            email = "newname@example.com"
        )

        sut.updateProfile(existingUser, updatedUser, USER_AGENT)

        assertVerificationEmail(mailerPort)
        val user = givenExistingUser().copy(
            email = "newname@example.com"
        )
        verify { userPort.update(user.id, user) }
    }

    @Test
    fun changePasswordTooShortBody() {
        val user = givenExistingUser()

        assertThatThrownBy { sut.changePassword(user, "secret") }
            .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun changePasswordBody() {
        val user = givenExistingUser()

        sut.changePassword(user, "secretlong")

        verify { userPort.updateCredentials(eq(user.id), any()) }
        verify { authorizationPort.deleteAllByUser(user.name) }
    }

    @Test
    fun verifyEmailSuccess() {
        val token = "verification"
        val user = createNewUser().copy(
            id = EXISTING_USER_ID,
            emailVerification = token
        )
        every { userPort.findByEmailVerification(token) } returns user

        sut.emailVerification(token)

        verify {
            monitorPort.sendMessage(
                "Email verified {nickname='${USER_NAME}', email='${USER_EMAIL}'}"
            )
        }
        verify { userPort.updateEmailVerification(EXISTING_USER_ID, User.EMAIL_VERIFIED) }
    }

    @Test
    fun verifyEmailFailed() {
        val token = "verification"
        val user = createNewUser().copy(
            id = EXISTING_USER_ID,
            emailVerification = token
        )
        every { userPort.findByEmailVerification(token) } returns user

        sut.emailVerification("wrong_token")

        verify(exactly = 0) { monitorPort.sendMessage(any()) }
        verify(exactly = 0) {
            userPort.updateEmailVerification(
                EXISTING_USER_ID,
                User.EMAIL_VERIFIED
            )
        }
    }

    @Test
    fun resendEmailVerification() {
        val user = givenExistingUser()

        sut.resendEmailVerification(user)

        assertVerificationEmail(mailerPort)
        verify { userPort.updateEmailVerification(eq(EXISTING_USER_ID), any()) }
    }

    @Test
    fun deleteMyProfile() {
        val user = givenExistingUser()

        sut.deleteProfile(user, USER_AGENT)

        verify { userPort.anonymizeUser(EXISTING_USER_ID) }
        verify { userPort.addUsernameToBlocklist(USER_NAME) }
        verify { authorizationPort.deleteAllByUser(USER_NAME) }
    }

    @Test
    fun registerUserNameTaken() {
        val user = givenExistingUser()
        val newUser = createNewUser().copy(
            name = user.name
        )

        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(ProfileConflictException::class.java)
    }

    @Test
    fun registerUserNameBlocked() {
        val newUser = createNewUser().copy(
            name = "Blocked Name"
        )
        every { userPort.countBlockedUsername("blockedname") } returns 1

        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(ProfileConflictException::class.java)
    }

    @Test
    fun registerUserEmailTaken() {
        val user = givenExistingUser()
        val newUser = createNewUser().copy(
            name = "othername",
            email = user.email
        )

        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(ProfileConflictException::class.java)

        verify { monitorPort.sendMessage("Registration for user 'othername' with eMail 'existing@example.com' failed, eMail is already taken\nvia UserAgent") }
    }

    @Test
    fun registernUserNameTaken() {
        val user = givenExistingUser()
        val newUser = createNewUser().copy(
            name = user.name,
            email = "otheremail@example.com"
        )

        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(ProfileConflictException::class.java)

        verify {
            monitorPort.sendMessage(
                "Registration for user '%s' with eMail '%s' failed, name is already taken by different eMail '%s'%nvia %s"
                    .format(user.name, newUser.email, user.email, USER_AGENT)
            )
        }
    }

    @Test
    fun registerUserWithEmptyName() {
        val newUser = createNewUser().copy(
            name = ""
        )

        assertThatThrownBy { sut.register(newUser, USER_AGENT) }
            .isInstanceOf(IllegalArgumentException::class.java)
    }

    private fun assertNewPasswordEmail() {
        verify(exactly = 1) {
            mailerPort.send(
                any(),
                any(), match {
                    it.matches(
                        """
                        Hello,
                        
                        your new password is: .*
                        
                        Cheers
                        Your Railway-Stations-Team
                        """.trimIndent().toRegex()
                    )
                })
        }
    }

    private fun createNewUser(): User {
        val key =
            "246172676F6E32696424763D3139246D3D36353533362C743D322C703D3124426D4F637165757646794E44754132726B566A6A3177246A7568362F6E6C2F49437A4B475570446E6B674171754A304F7A486A62694F587442542F2B62584D49476300000000000000000000000000000000000000000000000000000000000000"
        return User(
            name = USER_NAME,
            url = "https://link@example.com",
            license = License.CC0_10,
            email = USER_EMAIL,
            ownPhotos = true,
            key = key,
        )
    }

    private fun givenExistingUser(): User {
        val user = createNewUser().copy(
            id = EXISTING_USER_ID
        )
        every { userPort.findByEmail(user.email!!) } returns user
        every { userPort.findByNormalizedName(user.name) } returns user
        return user
    }

    @Test
    fun resetPasswordUnknownUser() {
        assertThatThrownBy { sut.resetPassword("unknown_user", USER_AGENT) }
            .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun resetPasswordEmptyEmail() {
        every { userPort.findByNormalizedName(USER_NAME) } returns createNewUser().copy(email = null)

        assertThatThrownBy { sut.resetPassword(USER_NAME, USER_AGENT) }
            .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun resetPasswordViaUsernameEmailNotVerified() {
        val user = createNewUser().copy(
            id = 123
        )
        every { userPort.findByNormalizedName(USER_NAME) } returns user

        sut.resetPassword(USER_NAME, USER_AGENT)

        verify { userPort.updateCredentials(eq(123), any()) }
        verify {
            monitorPort.sendMessage(
                "Reset Password for '%s', email='%s'".format(
                    USER_NAME, USER_EMAIL
                )
            )
        }
        assertNewPasswordEmail()
        verify { userPort.updateEmailVerification(123, User.EMAIL_VERIFIED_AT_NEXT_LOGIN) }
        verify { authorizationPort.deleteAllByUser(USER_NAME) }
    }

    @Test
    fun resetPasswordViaEmailAndEmailVerified() {
        val user = createNewUser().copy(
            id = 123,
            emailVerification = User.EMAIL_VERIFIED

        )
        every { userPort.findByEmail(user.email!!) } returns user

        sut.resetPassword(user.email!!, USER_AGENT)

        verify { userPort.updateCredentials(eq(123), any()) }
        verify {
            monitorPort.sendMessage(
                "Reset Password for '%s', email='%s'".format(
                    USER_NAME, USER_EMAIL
                )
            )
        }
        assertNewPasswordEmail()
        verify(exactly = 0) { userPort.updateEmailVerification(123, User.EMAIL_VERIFIED_AT_NEXT_LOGIN) }
        verify { authorizationPort.deleteAllByUser(user.name) }
    }

    fun assertVerificationEmail(mailerPort: MailerPort) {
        verify(exactly = 0) {
            mailerPort.send(
                any(),
                any(), match {
                    it.matches(
                        """
                                Hello,

                                please click on EMAIL_VERIFICATION_URL.* to verify your eMail-Address.

                                Cheers
                                Your Railway-Stations-Team

                                ---
                                Hallo,

                                bitte klicke auf EMAIL_VERIFICATION_URL.*, um Deine eMail-Adresse zu verifizieren.

                                Viele Grüße
                                Dein Bahnhofsfoto-Team
                                """.trimIndent().toRegex()
                    )
                }
            )
        }
    }
}